### PR TITLE
cmake: format and reorder cmake file, add uninstall target

### PR DESCRIPTION
**Pull request by dirkolbrich**
This PR formats and reorders the current cmake file:

- format cmake to use consistent lower command capitalization and indentation, as per the Cmake Coding Style, see https://community.kde.org/Policies/CMake_Coding_Style
- make search for Autodesk 3ds SDK optional, will only be performed when the option BUILD_IFCMAX is set
- add an uninstall target, see the Cmake FAQ https://gitlab.kitware.com/cmake/community/-/wikis/FAQ#can-i-do-make-uninstall-with-cmake
- minor reordering of code block to make the cmake file better readable
- fix warning for mismatching arguments

Fix the minor spelling error of using upper case `OR` in opening if else block and ending with lower case `or` in closing block. This gave the following error during build:

```
CMake Warning (dev) in CMakeLists.txt:
  A logical block opening on the line

    ~/IfcOpenShell/cmake/CMakeLists.txt:775 (if)

  closes on the line

    ~/IfcOpenShell/cmake/CMakeLists.txt:796 (endif)

  with mis-matching arguments.
This warning is for project developers.  Use -Wno-dev to suppress it.
```

### CODE REVIEW

**Comment by dirkolbrich** [comment]
I also formatted and unified the complete CMakeLists.txt file:
- consistent use of command capitalization
- indentation

**Comment by dirkolbrich** [comment]
ToDo: find a way to `make uninstall` the empty folders as well. Though some argue to better leave folders untouched as not to accidentally delete other files installed by other processes.

**Comment by dirkolbrich** [comment]
Bump this topic. Is there any interest in this project for improvement of code style uniformity or improvement of the Cmake setup?

**Comment by dirkolbrich** [comment]
Hey another bump for this. I would like to support this project and maybe start in making the CMake files a bit more modular. The current CMake file has > 1000 LOC and is IMV not well ordered in a logically way.

- unify cmake formating (see above commits)
- general ordering in logical chunks
- split up this huge CMake file and maybe move configuration of external dependencies into sub-files, e.g. ConfigureOCC.txt, ConfiqureBoost.txt etc.
- move the main CMakeList.txt. file under /root as it is general standard in other projects

Right now if I try to rebase my fix branch on head (branch v0.7.0) I get quite a bit of merge conflicts. I would appreciate some help on this. What can I do to move this forward?

**Comment by Moult** [comment]
Bump @aothms :) this PR was mentioned here https://community.osarch.org/discussion/comment/17100#Comment_17100

**Comment by dirkolbrich** [comment]
I think I fixed most of the merge conflict. Please have a look.
Next step would be splitting this huge CMakeLists.file into smaller components and moving inline function/macros into separate files.

**Comment by aothms** [comment]
So sorry for the delay. As said on various occasions I really wanted to get v0.8 out and rebase most of the PRs, but it proved to be too much of a monumental effort.

For reference here is a diff of this PR with `-w -B --ignore-case` to see the more functional differences.

```diff
--- v0.7.0:cmake/CMakeLists.txt     2023-08-30 10:38:32.122418200 +0200
+++ dirkolbrich/fix-cmake-mismatch-argument:cmake/CMakeLists.txt      2023-08-30 10:39:08.137699900 +0200
@@ -30,2 +30,6 @@

+if(NOT CMAKE_BUILD_TYPE)
+    set(CMAKE_BUILD_TYPE "Release")
+endif()
+
 # use extra version to make pre-release using eg semver
@@ -33,9 +37,2 @@

-foreach(max_year RANGE 2014 2030)
-set(max_sdk "$ENV{ADSK_3DSMAX_SDK_${max_year}}")
-if (NOT "${max_sdk}" STREQUAL "")
-MESSAGE(STATUS "Autodesk 3ds Max SDK found at ${max_sdk}")
-set(HAS_MAX TRUE)
-endif()
-endforeach()
 OPTION(MINIMAL_BUILD "The build is to make a minimal version of IFC converter from OCCT into IFC." OFF)
@@ -58,5 +56,3 @@
 OPTION(WASM_BUILD OFF)
-if (${HAS_MAX})
-OPTION(BUILD_IFCMAX "Build IfcMax, a 3ds Max plug-in, Windows-only." OFF)
-endif()
+
 if (${BUILD_CONVERT})
@@ -67,6 +64,5 @@
 OPTION(ADD_COMMIT_SHA "Add commit sha and branch in version number, warning results in many rebuilds, requires git" OFF)
-
-IF(NOT CMAKE_BUILD_TYPE)
-       SET(CMAKE_BUILD_TYPE "Release")
-ENDIF()
+option(BUILD_IFCMAX "Build IfcMax, a 3ds Max plug-in, Windows-only." OFF)
+# QtViewer requires Qt5
+option(BUILD_QTVIEWER "Build IfcOpenShell Qt GUI Viewer" OFF)

@@ -84,4 +80,2 @@

-# QtViewer requires Qt5
-OPTION(BUILD_QTVIEWER "Build IfcOpenShell Qt GUI Viewer" OFF)
 include(GNUInstallDirs)
@@ -139,2 +143,5 @@
 UNIFY_ENVVARS_AND_CACHE(OCC_LIBRARY_DIR)
+UNIFY_ENVVARS_AND_CACHE(BOOST_ROOT)
+UNIFY_ENVVARS_AND_CACHE(BOOST_LIBRARYDIR)
+
 if (NOT MINIMAL_BUILD)
@@ -156,4 +163,2 @@
 endif()
-UNIFY_ENVVARS_AND_CACHE(BOOST_ROOT)
-UNIFY_ENVVARS_AND_CACHE(BOOST_LIBRARYDIR)

@@ -415,6 +436,2 @@

-if(WASM_BUILD)
-    set(CMAKE_FIND_ROOT_PATH "${CMAKE_FIND_ROOT_PATH_BACKUP}")
-endif()
-
 if(OCCT_STATIC)
@@ -712,4 +741,4 @@

-foreach(s ${SCHEMA_VERSIONS})
-       add_definitions(-DHAS_SCHEMA_${s})
+foreach(schema ${SCHEMA_VERSIONS})
+    add_definitions(-DHAS_SCHEMA_${schema})
 endforeach()
@@ -769,5 +803,5 @@
     # Prevent the schema that had just been compiled from being excluded
-       foreach(s ${SCHEMA_VERSIONS})
-               if("${COMPILED_SCHEMA_NAME}" STREQUAL "${s}")
-                       list(REMOVE_ITEM IFC_RELEASE_NOT_USED "${s}")
+    foreach(schema ${SCHEMA_VERSIONS})
+        if("${COMPILED_SCHEMA_NAME}" STREQUAL "${schema}")
+            list(REMOVE_ITEM IFC_RELEASE_NOT_USED "${schema}")
                endif()
@@ -783,4 +817,4 @@
 if (BUILD_IFCGEOM)
-       foreach(s ${SCHEMA_VERSIONS})
-               set(IFCGEOM_SCHEMA_LIBRARIES ${IFCGEOM_SCHEMA_LIBRARIES} IfcGeom_ifc${s})
+       foreach(schema ${SCHEMA_VERSIONS})
+               set(IFCGEOM_SCHEMA_LIBRARIES ${IFCGEOM_SCHEMA_LIBRARIES} IfcGeom_ifc${schema})
        endforeach()
@@ -793,4 +828,4 @@
 if (BUILD_CONVERT OR BUILD_IFCPYTHON)
-       foreach(s ${SCHEMA_VERSIONS})
-               set(SERIALIZER_SCHEMA_LIBRARIES ${SERIALIZER_SCHEMA_LIBRARIES} Serializers_ifc${s})
+       foreach(schema ${SCHEMA_VERSIONS})
+               set(SERIALIZER_SCHEMA_LIBRARIES ${SERIALIZER_SCHEMA_LIBRARIES} Serializers_ifc${schema})
        endforeach()
@@ -803,6 +838,7 @@

-foreach(s ${IFCPARSE_H_FILES_ALL})
-get_filename_component(p "${s}" NAME)
-if (NOT "${p}" MATCHES "[0-9]")
-list(APPEND IFCPARSE_H_FILES "${s}")
+foreach(file ${IFCPARSE_H_FILES_ALL})
+    get_filename_component(filename "${file}" NAME)
+
+    if(NOT "${filename}" MATCHES "[0-9]")
+        list(APPEND IFCPARSE_H_FILES "${file}")
 endif()
@@ -810,6 +846,7 @@

-foreach(s ${IFCPARSE_CPP_FILES_ALL})
-get_filename_component(p "${s}" NAME)
-if (NOT "${p}" MATCHES "[0-9]")
-list(APPEND IFCPARSE_CPP_FILES "${s}")
+foreach(file ${IFCPARSE_CPP_FILES_ALL})
+    get_filename_component(filename "${file}" NAME)
+
+    if(NOT "${filename}" MATCHES "[0-9]")
+        list(APPEND IFCPARSE_CPP_FILES "${file}")
 endif()
@@ -817,10 +854,10 @@

-foreach(s ${SCHEMA_VERSIONS})
+foreach(schema ${SCHEMA_VERSIONS})
        list(APPEND IFCPARSE_H_FILES
-               ../src/ifcparse/Ifc${s}.h
-               ../src/ifcparse/Ifc${s}-definitions.h
+        ../src/ifcparse/Ifc${schema}.h
+        ../src/ifcparse/Ifc${schema}-definitions.h
        )
        list(APPEND IFCPARSE_CPP_FILES
-               ../src/ifcparse/Ifc${s}.cpp
-               ../src/ifcparse/Ifc${s}-schema.cpp
+        ../src/ifcparse/Ifc${schema}.cpp
+        ../src/ifcparse/Ifc${schema}-schema.cpp
        )
@@ -840,4 +876,4 @@

-if (BUILD_IFCGEOM)

+if(BUILD_IFCGEOM)
 # IfcGeom
@@ -847,7 +883,7 @@

-foreach(s ${SCHEMA_VERSIONS})
-       add_library(IfcGeom_ifc${s} STATIC ${IFCGEOM_FILES})
-       set_target_properties(IfcGeom_ifc${s} PROPERTIES COMPILE_FLAGS "-DIFC_GEOM_EXPORTS -DIfcSchema=Ifc${s}")
+foreach(schema ${SCHEMA_VERSIONS})
+       add_library(IfcGeom_ifc${schema} STATIC ${IFCGEOM_FILES})
+       set_target_properties(IfcGeom_ifc${schema} PROPERTIES COMPILE_FLAGS "-DIFC_GEOM_EXPORTS -DIfcSchema=Ifc${schema}")
        if (NOT WASM_BUILD)
-           TARGET_LINK_LIBRARIES(IfcGeom_ifc${s} IfcParse ${OPENCASCADE_LIBRARIES})
+        target_link_libraries(IfcGeom_ifc${schema} IfcParse ${OPENCASCADE_LIBRARIES})
        endif()
@@ -885,10 +920,10 @@

-foreach(s ${SCHEMA_VERSIONS})
-       add_library(Serializers_ifc${s} STATIC ${SERIALIZERS_S_FILES})
-       set_target_properties(Serializers_ifc${s} PROPERTIES COMPILE_FLAGS "-DIFC_GEOM_EXPORTS -DIfcSchema=Ifc${s} ${CONVERT_PRECISION}")
+foreach(schema ${SCHEMA_VERSIONS})
+       add_library(Serializers_ifc${schema} STATIC ${SERIALIZERS_S_FILES})
+       set_target_properties(Serializers_ifc${schema} PROPERTIES COMPILE_FLAGS "-DIFC_GEOM_EXPORTS -DIfcSchema=Ifc${schema} ${CONVERT_PRECISION}")

        if (WASM_BUILD)
-           TARGET_LINK_LIBRARIES(Serializers_ifc${s} ${HDF5_LIBRARIES})
+        target_link_libraries(Serializers_ifc${schema} ${HDF5_LIBRARIES})
        else()
-           TARGET_LINK_LIBRARIES(Serializers_ifc${s} IfcGeom ${OPENCASCADE_LIBRARIES} ${HDF5_LIBRARIES})
+        target_link_libraries(Serializers_ifc${schema} IfcGeom ${OPENCASCADE_LIBRARIES} ${HDF5_LIBRARIES})
        endif()
@@ -955,3 +987,3 @@
        execute_process(
-               COMMAND ${GIT_EXECUTABLE} branch -a --contains HEAD
+            COMMAND ${GIT_EXECUTABLE} branch --contains HEAD
                OUTPUT_VARIABLE git_branches
@@ -961,7 +994,6 @@
        foreach(git_branch_candidate IN ITEMS ${git_branch_list})
-               string(REPLACE "*" "" git_branch_candidate_temp "${git_branch_candidate}")
-               string(STRIP "${git_branch_candidate_temp}" git_branch_candidate_2)
+            string(STRIP "${git_branch_candidate}" git_branch_candidate_2)
+
                if (NOT git_branch_candidate_2 MATCHES "^HEAD$")
-                       string(REPLACE "/" ";" git_branch_candidate_2_list "${git_branch_candidate_2}")
-                       list(GET git_branch_candidate_2_list -1 git_branch)
+                set(git_branch ${git_branch_candidate_2})
                endif()
@@ -995,3 +1028,16 @@
 IF(NOT MINIMAL_BUILD AND BUILD_IFCMAX)
+    foreach(max_year RANGE 2014 2030)
+        set(max_sdk "$ENV{ADSK_3DSMAX_SDK_${max_year}}")
+
+        if(NOT "${max_sdk}" STREQUAL "")
+            message(STATUS "Autodesk 3ds Max SDK found at ${max_sdk}")
+            set(HAS_MAX TRUE)
+        endif()
+    endforeach()
+
+    if(HAS_MAX)
        ADD_SUBDIRECTORY(../src/ifcmax ifcmax)
+    else()
+        message(STATUS "Autodesk 3ds Max SDK not found, is required to build IFCMax.")
+    endif()
 ENDIF()
@@ -1001,2 +1048,6 @@

+if(BUILD_QTVIEWER)
+    add_subdirectory(../src/qtviewer qtviewer)
+endif()
+
 # CMake installation targets
@@ -1044,6 +1095,14 @@

-IF(BUILD_QTVIEWER)
-    ADD_SUBDIRECTORY(../src/qtviewer qtviewer)
+# Cmake uninstall target
+if(NOT TARGET uninstall)
+    configure_file(
+        "${CMAKE_CURRENT_SOURCE_DIR}/cmake_uninstall.cmake.in"
+        "${CMAKE_CURRENT_BINARY_DIR}/cmake_uninstall.cmake"
+        IMMEDIATE @ONLY)
+
+    add_custom_target(uninstall
+        COMMAND ${CMAKE_COMMAND} -P ${CMAKE_CURRENT_BINARY_DIR}/cmake_uninstall.cmake)
 endif()

+# Packaging
 list(APPEND CPACK_SOURCE_IGNORE_FILES
@@ -1083,4 +1140,4 @@
 set(CPACK_DEBIAN_ARCHITECTURE "${CMAKE_SYSTEM_PROCESSOR}")
-#set(CPACK_DEBIAN_PACKAGE_CONTROL_EXTRA "${CMAKE_SOURCE_DIR}/cmake/debian/postinst")

+# set(CPACK_DEBIAN_PACKAGE_CONTROL_EXTRA "${CMAKE_SOURCE_DIR}/cmake/debian/postinst")
 include(CPack)
```

**Comment by aothms** [comment]
Based on the diff above I have the following comments:

```diff
-if(WASM_BUILD)
-    set(CMAKE_FIND_ROOT_PATH "${CMAKE_FIND_ROOT_PATH_BACKUP}")
-endif()
```

^ Probably rightfully removed. Was a leftover of using a cross compilation sysroot trying to get wasm up and running.

```diff
-               COMMAND ${GIT_EXECUTABLE} branch -a --contains HEAD
+            COMMAND ${GIT_EXECUTABLE} branch --contains HEAD
```

^ Not sure about this though is the removal of `-a` for a reason? If I recall correctly it was needed to get proper branch reported on the daily builds on OSX.

```diff
-               string(REPLACE "*" "" git_branch_candidate_temp "${git_branch_candidate}")
-               string(STRIP "${git_branch_candidate_temp}" git_branch_candidate_2)
+            string(STRIP "${git_branch_candidate}" git_branch_candidate_2)
```

^ Same for the replace of `"*"` to `""`. Likely a merge conflict resolution error?

```diff
-                       string(REPLACE "/" ";" git_branch_candidate_2_list "${git_branch_candidate_2}")
-                       list(GET git_branch_candidate_2_list -1 git_branch)
+                set(git_branch ${git_branch_candidate_2})
```

^ Same?

**Comment by dirkolbrich** [comment]
1. WASM_BUILD
> Based on the diff above I have the following comments:
```diff
-if(WASM_BUILD)
-    set(CMAKE_FIND_ROOT_PATH "${CMAKE_FIND_ROOT_PATH_BACKUP}")
-endif()
```
> ^ Probably rightfully removed. Was a leftover of using a cross compilation sysroot trying to get wasm up and running.

I see I deleted a reset of `CMAKE_FIND_ROOT_PATH`, which was early set to `""`, see [line 160 - line 166](https://github.com/IfcOpenShell/IfcOpenShell/blob/fbd8ea1edb383ffcd07ab319996e4e27d1790e9c/cmake/CMakeLists.txt#L160-166)

```cmake
if(WASM_BUILD)
    # when using the nix/build-all.py build script we should not
    # look into the sysroot for most of the dependencies but rather
    # in the designated build/ folder created by the script.
    set(CMAKE_FIND_ROOT_PATH_BACKUP "${CMAKE_FIND_ROOT_PATH}")
    set(CMAKE_FIND_ROOT_PATH "")
endif()
```

It is not entirely clear which parts of the `CMakeLists.txt` file should use the empty `CMAKE_FIND_ROOT_PATH` string and which not. Can this setting and resetting deleted entirely or should I copy the reset back in?

**Comment by dirkolbrich** [comment]
2. Git setup

```diff
-               COMMAND ${GIT_EXECUTABLE} branch -a --contains HEAD
+            COMMAND ${GIT_EXECUTABLE} branch --contains HEAD
```
> ^ Not sure about this though is the removal of -a for a reason? If I recall correctly it was needed to get proper branch reported on the daily builds on OSX.
```diff
-               string(REPLACE "*" "" git_branch_candidate_temp "${git_branch_candidate}")
-               string(STRIP "${git_branch_candidate_temp}" git_branch_candidate_2)
+            string(STRIP "${git_branch_candidate}" git_branch_candidate_2)
```
> ^ Same for the replace of "*" to "". Likely a merge conflict resolution error?
```diff
-                       string(REPLACE "/" ";" git_branch_candidate_2_list "${git_branch_candidate_2}")
-                       list(GET git_branch_candidate_2_list -1 git_branch)
+                set(git_branch ${git_branch_candidate_2})
```
> ^ Same?

Yes, my error. This is a mistake by trying to solve the merge conflicts. Will copy the original code back in.

**Comment by dirkolbrich** [comment]
Thought about the `WASM_BUILD` change. As this PR is mostly about style unification and some reordering, I will copy the reset if  `CMAKE_FIND_ROOT_PATH` back in. This change should be handled in a different commit entirely.

**Comment by aothms** [comment]
> It is not entirely clear which parts of the CMakeLists.txt file should use the empty CMAKE_FIND_ROOT_PATH string and which not. Can this setting and resetting deleted entirely or should I copy the reset back in?
> 
> Edit: The setting of CMAKE_FIND_ROOT_PATH is global, yet the reset is within a if(BUILD_IFCGEOM) condition. Is this intentional?

I don't think that's intentional. I really wish the approach in cmake was much more declarative instead of all this imperative if-else spaghetti. If you have any ideas regarding that, it'd be very welcome. 

The WASM_BUILD setting is only tested in conjunction with the nix build script. My guess: every dependencies built by the build script is supposed to be taken from the empty sysroot. Every other dependency taken from the emscriptem sysroot.

The list of dependencies built by the script is here:

https://github.com/IfcOpenShell/IfcOpenShell/blob/9cc1f5f0e3177e8181dd4a586f61d4aa7b525043/nix/build-all.py#L187-L204

I hope that gives you some context on how to proceed with the sysroot thing.

> This is a mistake by trying to solve the merge conflicts. Will copy the original code back in.

Thanks a lot and sorry for taking so much time resulting in these conflicts.

File: cmake/CMakeLists.txt [hunk 1]
```diff
@@ -575,118 +594,128 @@ if(NOT MINIMAL_BUILD AND HDF5_SUPPORT)
...
             /usr/lib/x86_64-linux-gnu/libaec.so
             z dl
         )
     endif()
-    
+
     add_definitions(-DWITH_HDF5)
-	set(SWIG_DEFINES ${SWIG_DEFINES} -DWITH_HDF5)
+    set(SWIG_DEFINES ${SWIG_DEFINES} -DWITH_HDF5)
 endif()
```
Comment (dirkolbrich) [inline cmake/CMakeLists.txt:619]: Preparation of external library paths is done here. Add reset of "CMAKE_FIND_ROOT_PATH" here back in, right before other actions like complier settings etc.

**Comment by dirkolbrich** [comment]
OK, thanks for the context. So I think the setting of `CMAKE_FIND_ROOT_PATH` to "" is only useful during the preparation of all dependent library path. The rest should occur after this is done and before other action. 

I would suggest adding it in right before other compiler settings, see my comment here: https://github.com/IfcOpenShell/IfcOpenShell/pull/2393/files#r1311338883

**Comment by aothms** [comment]
Looks good to me 👍

**Comment by aothms** [comment]
@dirkolbrich is the ready to be merged now from your side?

**Comment by dirkolbrich** [comment]
From my side yes. Tested a build on macOS M1 which works.
Fixing of deprecation warnings and other optimisations come in the next steps.

**Comment by aothms** [comment]
Awesome, thanks again for your patience and dedication to this :)

**Pull request merged**